File: SmartReceipts/Utils/Import/ImagePicker.m
//
//  ImagePicker.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 22/04/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "ImagePicker.h"
#import "Constants.h"
#import "WBImageUtils.h"
#import "WBPreferences.h"
#import "LocalizedString.h"

@interface ImagePicker () <UINavigationControllerDelegate, UIImagePickerControllerDelegate>

@property (nonatomic, copy) WBImagePickerResultBlock selectionHandler;

@end

@implementation ImagePicker

+ (instancetype)sharedInstance {
    DEFINE_SHARED_INSTANCE_USING_BLOCK(^{
        return [[self alloc] initSingleton];
    });
}

- (id)init {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:[NSString stringWithFormat:@"You must use [%@ %@] instead",
                                                                     NSStringFromClass([self class]),
                                                                     NSStringFromSelector(@selector(sharedInstance))]
                                 userInfo:nil];
    return nil;
}

- (id)initSingleton {
    self = [super init];
    if (self) {
        // Custom initialization code
    }
    return self;
}

- (void)presentPickerOnController:(UIViewController *)controller completion:(WBImagePickerResultBlock)completion {
    [self setSelectionHandler:completion];

    BOOL hasCamera = [UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera];
    if (!hasCamera) {
        [self presentImagePickerWithSource:UIImagePickerControllerSourceTypePhotoLibrary onController:controller];
        return;
    }
    
    NSString *title = LocalizedString(@"image.picker.sheet.title", nil);
    UIAlertControllerStyle style = UIAlertControllerStyleActionSheet;
    UIAlertController *actionSheet = [UIAlertController alertControllerWithTitle:title message:nil preferredStyle:style];
    
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:LocalizedString(@"DIALOG_CANCEL", nil)
        style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
             [self setSelectionHandler:nil];
        }];
    
    UIAlertAction *library = [UIAlertAction actionWithTitle:LocalizedString(@"image.picker.sheet.choose.existing.button.title", nil)
       style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
           [self presentImagePickerWithSource:UIImagePickerControllerSourceTypePhotoLibrary onController:controller];
       }];
    
    UIAlertAction *photo = [UIAlertAction actionWithTitle:LocalizedString(@"image.picker.sheet.take.photo.button.title", nil)
       style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
           [self presentImagePickerWithSource:UIImagePickerControllerSourceTypeCamera onController:controller];
       }];
    
    [actionSheet addAction:photo];
    [actionSheet addAction:library];
    [actionSheet addAction:cancel];
    
    [controller presentViewController:actionSheet animated:YES completion:nil];
}

- (void)presentCameraOnController:(UIViewController *)controller completion:(WBImagePickerResultBlock)completion {
    [self setSelectionHandler:completion];
    [self presentImagePickerWithSource:UIImagePickerControllerSourceTypeCamera onController:controller];
}

- (void)presentImagePickerWithSource:(UIImagePickerControllerSourceType)source onController:(UIViewController *)controller {
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.sourceType = source;

    [controller presentViewController:picker animated:YES completion:NULL];
}

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info {
    UIImage *chosenImage = info[UIImagePickerControllerOriginalImage];

    chosenImage = [WBImageUtils processImage:chosenImage];
    CGFloat compressionRation = 0.95;
    chosenImage = [WBImageUtils compressImage:chosenImage withRatio:compressionRation];

    [picker dismissViewControllerAnimated:YES completion:^{
        self.selectionHandler(chosenImage);
        [self setSelectionHandler:nil];
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    self.selectionHandler(nil);
    [self setSelectionHandler:nil];
    [picker dismissViewControllerAnimated:YES completion:NULL];
}

@end

File: SmartReceipts/Common/UI/ImagePicker.swift
//
//  ImagePicker.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 14/11/2018.
//  Copyright © 2018 Will Baumann. All rights reserved.
//

import UIKit
import RxSwift
import AVKit

private let COMPRESSION_RATIO: CGFloat = 0.95

class ImagePicker: NSObject {
    static let shared = ImagePicker()
    private override init() {}
    
    private var singleObserver: ((SingleEvent<UIImage>) -> ())?
    
    private func result() -> Single<UIImage> {
        return .create(subscribe: { observer -> Disposable in
            self.singleObserver = observer
            return Disposables.create()
        })
    }
    
    //MARK: - Interface
    
    func presentPicker(on viewController: UIViewController) -> Single<UIImage> {
        let hasCamera = UIImagePickerController.isSourceTypeAvailable(.camera)
        if !hasCamera {
            return presentPicker(on: viewController, source: .photoLibrary)
        }
        
        let actionSheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        
        let cancel = UIAlertAction(title: LocalizedString("DIALOG_CANCEL"), style: .cancel) { _ in
            self.singleObserver = nil
        }
        
        let library = UIAlertAction(title: LocalizedString("receipt_attach_image"), style: .default) { _ in
            self.presentPicker(on: viewController, source: .photoLibrary)
        }
        
        let photo = UIAlertAction(title: LocalizedString("receipt_attach_photo"), style: .default) { _ in
            self.presentPicker(on: viewController, source: .camera)
        }

        actionSheet.addAction(photo)
        actionSheet.addAction(library)
        actionSheet.addAction(cancel)
        
        viewController.present(actionSheet, animated: true, completion: nil)
        
        return result()
    }
    
    func presentCamera(on viewController: UIViewController) -> Single<UIImage> {
        return presentPicker(on: viewController, source: .camera)
    }
    
    @discardableResult
    private func presentPicker(on viewController: UIViewController, source: UIImagePickerController.SourceType) -> Single<UIImage> {
        let picker = UIImagePickerController()
        
        if AVCaptureDevice.authorizationStatus(for: .video) == .denied {
            let title = LocalizedString("generic_error_alert_title")
            let message = LocalizedString("camera_permission_alert_message")
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
            let cancel = UIAlertAction(title: LocalizedString("DIALOG_CANCEL"), style: .cancel, handler: nil)
            let settings = UIAlertAction(title: LocalizedString("camera_permission_alert_button"), style: .default) { _ in
                guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            }
            
            alert.addAction(cancel)
            alert.addAction(settings)
            
            viewController.present(alert, animated: true, completion: nil)
            return .never()
        }
        
        picker.delegate = self
        picker.sourceType = source
        
        viewController.present(picker, animated: true, completion: nil)
        
        return result()
    }
    
}

extension ImagePicker: UIImagePickerControllerDelegate, UINavigationControllerDelegate {

    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        var chosenImage = info[.originalImage] as! UIImage
        chosenImage = WBImageUtils.processImage(chosenImage)
        chosenImage = WBImageUtils.compressImage(chosenImage, withRatio: COMPRESSION_RATIO)
        
        
        picker.dismiss(animated: true) {
            self.singleObserver?(.success(chosenImage))
        }
    }
    
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true, completion: nil)
    }
    
}
